### PR TITLE
reveal-md: init at 5.2.0

### PR DESCRIPTION
**Pull request by sagikazarmark**
Signed-off-by: Mark Sagi-Kazar <mark.sagikazar@gmail.com>

###### Motivation for this change

Build reveal.js slides from markdown.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/node-packages/node-packages.json [hunk 1]
```diff
@@ -237,6 +237,7 @@
 , "redoc-cli"
 , "remod-cli"
 , "reveal.js"
+, "reveal-md"
```
Comment (jonringer) [inline pkgs/development/node-packages/node-packages.json:240]: you will probably want to add `inherit (nodePackages) reveal-md;` to `pkgs/top-level/all-packages.nix` for this to be easy-to-find

Comment (sagikazarmark) [reply pkgs/development/node-packages/node-packages.json:240]: Thank you! Fixed.

**Review by jonringer — CHANGES_REQUESTED** [review]
Getting a build failure:

```
ERROR: Failed to download Chromium r674921! Set "PUPPETEER_SKIP_CHROMIUM_DOWNLOAD" env variable to skip download.
Error: getaddrinfo ENOTFOUND storage.googleapis.com
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:71:26)
  -- ASYNC --
    at BrowserFetcher.<anonymous> (/nix/store/dqxfrsbgq0v7wscry5s4pl1cd6wdbkx1-reveal-md-5.2.0/lib/node_modules/reveal-md/n>
    at Object.<anonymous> (/nix/store/dqxfrsbgq0v7wscry5s4pl1cd6wdbkx1-reveal-md-5.2.0/lib/node_modules/reveal-md/node_modu>
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
    at internal/main/run_main_module.js:17:47 {
  errno: -3008,
  code: 'ENOTFOUND',
  syscall: 'getaddrinfo',
  hostname: 'storage.googleapis.com'
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! puppeteer@1.19.0 install: `node install.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the puppeteer@1.19.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

you will probably need to do something to:
```
  reveal-md = super.reveal-md.override({
    prePatch = ''
      export PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=1
    '';
  });
```
in `pkgs/development/node-packages/default.nix`

**Comment by sagikazarmark** [comment]
@jonringer interesting, thanks for pointing that out.

I looked at other examples and went with an existing solution that the mermaid-js package uses as well.
Chromium is not supported on Darwin, so I disabled downloading on every other platform.

Let me know if it's not the right solution.

Thanks again!

File: pkgs/development/node-packages/default.nix [hunk 2]
```diff
@@ -340,6 +340,20 @@ let
       '';
     };
 
+    reveal-md = super.reveal-md.override (
+      if stdenv.isDarwin
+      then {}
+      else {
```
Comment (jonringer) [inline pkgs/development/node-packages/default.nix:346]:
```suggestion
      lib.optionalAttrs (!stdenv.isDarwin) {
```

Comment (sagikazarmark) [reply pkgs/development/node-packages/default.nix]: Done

File: pkgs/development/node-packages/default.nix [hunk 3]
```diff
@@ -340,6 +340,19 @@ let
       '';
     };
 
+    reveal-md = super.reveal-md.override (
+      lib.optionalAttrs (!stdenv.isDarwin) {
```
Comment (SuperSandro2000) [inline pkgs/development/node-packages/default.nix:344]: What happens if I try to build the package on Darwin?

Comment (sagikazarmark) [reply pkgs/development/node-packages/default.nix:344]: It builds successfully for me. I found this pattern for other node apps based on puppeteer in default.nix

I don't know why it doesn't work on linux, but on darwin we can't use the chromium package.

Comment (sagikazarmark) [reply pkgs/development/node-packages/default.nix:344]: See [mermaid-cli](https://github.com/NixOS/nixpkgs/blob/5775aa907655ce9ee2a2e6470a9b3a124dcc3351/pkgs/development/node-packages/default.nix#L257)

**Comment by jonringer** [comment]
and the node conflicts begin :(

```
CONFLICT (content): Merge conflict in pkgs/development/node-packages/node-packages.nix
Auto-merging pkgs/development/node-packages/node-packages.json
Automatic merge failed; fix conflicts and then commit the result.
https://github.com/NixOS/nixpkgs/pull/144918 failed to build
```

**Comment by sagikazarmark** [comment]
I'm already rebasing it for like the 10th time now to resolve conflicts, but running the generate script takes like 2 hours every time. :)

**Comment by SuperSandro2000** [comment]
> I'm already rebasing it for like the 10th time now to resolve conflicts, but running the generate script takes like 2 hours every time. :)

We can also be sneaky and merge it without updating that file and let someone else do it.

**Comment by sagikazarmark** [comment]
@jonringer @SuperSandro2000 Rebase finally completed.

**Comment by jonringer** [comment]
other updates to node packages keep causing regressions....

**Comment by jonringer** [comment]
unrelated regression fix: #146166

**Review by jonringer — APPROVED** [review]
LGTM

failures are broken on target branch
```
https://github.com/NixOS/nixpkgs/pull/144918

1 package failed to build:
teck-programmer

4 packages built:
gtop lumo pm2 reveal-md
```

**Comment by sagikazarmark** [comment]
Thanks a lot for your help to get this merged!

**Comment by jonringer** [comment]
yea. node package within nix is not the greatest maintainer experience

**Pull request merged**